packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx

```diff
@@ -207,17 +210,19 @@ export const Datagrid: FC<DatagridProps> = React.forwardRef((props, ref) => {
         );
     }
 
+    const shouldRenderEmptyPage =
+        (data == null || data.length === 0 || total === 0) &&
+        filterValues != null && // Can be null when outside a <ListContext>
+        !Object.keys(filterValues).length &&
+        empty !== false;
+
```

Comment:
I added the check for `null` on the `filterValues` to make this works with the `<ArrayField>`. However, I'm not sure about what we should do here. It may be better to display the `empty` in `ArrayField` as well.